# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/html/HTMLMediaElement.cpp
# Commit: dea9f12cdec0
# Full Hash: dea9f12cdec02aeb5ac260a970659e09dc73a375
# Author: Andreas Pehrson <apehrson@mozilla.com>
# Date: 2025-07-02 09:15:46
# Regressor Bug: 1560936
# File Overlap Count: 3
# Description:
#   Bug 1560936 - For getSettings make all MediaStreamTrackSources populate width and height as appropriate. r=jib,bwc
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D251430
# ==============================================================================

diff -r a4ae1edb514a -r dea9f12cdec0 dom/html/HTMLMediaElement.cpp
--- a/dom/html/HTMLMediaElement.cpp	Tue Jul 01 18:42:42 2025 +0000
+++ b/dom/html/HTMLMediaElement.cpp	Tue Jul 01 18:42:43 2025 +0000
@@ -1123,13 +1123,15 @@
                                            MediaStreamTrackSource)
 
   /* MediaDecoder track source */
-  MediaElementTrackSource(ProcessedMediaTrack* aTrack, nsIPrincipal* aPrincipal,
-                          OutputMuteState aMuteState, bool aHasAlpha)
+  MediaElementTrackSource(HTMLMediaElement* aOwner, ProcessedMediaTrack* aTrack,
+                          nsIPrincipal* aPrincipal, OutputMuteState aMuteState,
+                          bool aHasAlpha)
       : MediaStreamTrackSource(
             aPrincipal, nsString(),
             TrackingId(TrackingId::Source::MediaElementDecoder,
                        sDecoderCaptureSourceId++,
                        TrackingId::TrackAcrossProcesses::Yes)),
+        mOwner(aOwner),
         mTrack(aTrack),
         mIntendedElementMuteState(aMuteState),
         mElementMuteState(aMuteState),
@@ -1138,7 +1140,8 @@
   }
 
   /* MediaStream track source */
-  MediaElementTrackSource(MediaStreamTrack* aCapturedTrack,
+  MediaElementTrackSource(HTMLMediaElement* aOwner,
+                          MediaStreamTrack* aCapturedTrack,
                           MediaStreamTrackSource* aCapturedTrackSource,
                           ProcessedMediaTrack* aTrack, MediaInputPort* aPort,
                           OutputMuteState aMuteState)
@@ -1147,6 +1150,7 @@
             TrackingId(TrackingId::Source::MediaElementStream,
                        sStreamCaptureSourceId++,
                        TrackingId::TrackAcrossProcesses::Yes)),
+        mOwner(aOwner),
         mCapturedTrack(aCapturedTrack),
         mCapturedTrackSource(aCapturedTrackSource),
         mTrack(aTrack),
@@ -1270,11 +1274,26 @@
     return mMediaDecoderHasAlpha.valueOr(false);
   }
 
+  void GetSettings(dom::MediaTrackSettings& aResult) override {
+    if (!mOwner) {
+      return;
+    }
+
+    auto* elem = mOwner->AsHTMLVideoElement();
+    if (!elem) {
+      return;
+    }
+
+    aResult.mWidth.Construct(elem->VideoWidth());
+    aResult.mHeight.Construct(elem->VideoHeight());
+  }
+
   ProcessedMediaTrack* Track() const { return mTrack; }
 
  private:
   virtual ~MediaElementTrackSource() { Destroy(); };
 
+  WeakPtr<HTMLMediaElement> mOwner;
   RefPtr<MediaStreamTrack> mCapturedTrack;
   RefPtr<MediaStreamTrackSource> mCapturedTrackSource;
   const RefPtr<ProcessedMediaTrack> mTrack;
@@ -3987,7 +4006,7 @@
         principal = NodePrincipal();
       }
       source = MakeAndAddRef<MediaElementTrackSource>(
-          track, principal, OutputTracksMuted(),
+          this, track, principal, OutputTracksMuted(),
           type == MediaSegment::VIDEO
               ? HTMLVideoElement::FromNode(this)->HasAlpha()
               : false);
@@ -4011,7 +4030,7 @@
       track = inputTrack->Graph()->CreateForwardedInputTrack(type);
       RefPtr<MediaInputPort> port = inputTrack->ForwardTrackContentsTo(track);
       source = MakeAndAddRef<MediaElementTrackSource>(
-          inputTrack, &inputTrack->GetSource(), track, port,
+          this, inputTrack, &inputTrack->GetSource(), track, port,
           OutputTracksMuted());
 
       // Track is muted initially, so we don't leak data if it's added while